src/parse.rs

```diff
@@ -0,0 +1,320 @@
+    PUSH_INT(i32),
+    POP,
+    ADD,
+    SUB,
+    LOAD_VAR(String),
+    STORE_VAR(String),
+    NEW_OBJECT,
+    LOAD_FIELD(String),
+    STORE_FIELD(String),
+    CLASS_LABEL(String),
+    METH_LABEL(String),
+    SWAP,
+    DUP,
+    CALL(i32),
+    JEQ(i32),
+    RET,
+}
+#[derive(Debug)]
+struct Class {
+    methods: HashMap<String, Method>,
+}
+
+impl Class {
+    fn new() -> Class {
+        Class {methods: HashMap::new()}
+    }
+
+    fn add_method(&mut self, name: String, method: Method) {
+        self.methods.insert(name, method);
+    }
+}
+
+
+#[derive(Debug)]
+struct Method {
+    locals: Vec<String>,
+    code:   Vec<Instr>
+}
+
+impl Method {
+    fn new() -> Method {
+        Method{
+            locals: Vec::new(),
+            code: Vec::new()
+        }
+    }
+
+    fn push_instr(&mut self, instr: Instr) {
+        self.code.push(instr)
+    }
+
+    fn push_var(&mut self, var_name: String) {
+        self.locals.push(var_name)
+    }
+
+}
+
+
+#[derive(Debug)]
+pub struct CompilerContext {
+    classes: HashMap<String,Class>,
+}
+
+impl CompilerContext {
+    fn add(&mut self, name: String, class: Class) {
+        self.classes.insert(name, class);
+    }
+}
+
+fn gen_bytecode(parse_tree: &Node<u16>, grm: &YaccGrammar, input: &str) -> CompilerContext {
+
+    fn id_value (node: &Node<u16>, grm: &YaccGrammar, input: &str) -> String {
+        if let &Node::Term { lexeme } = node {
+            return String::from(&input[lexeme.start()..lexeme.start() + lexeme.len()])
+        }
+        else { panic!("Identifier not found") }
+    };
+
+    fn term_name (node: &Node<u16>, grm: &YaccGrammar, input: &str) -> String {
+        match *node {
+            Node::Nonterm { nonterm_idx, ref nodes } => {
+                grm.nonterm_name(nonterm_idx).to_string()
+            }
+            Node::Term { lexeme } => {
+                let token_id: usize = lexeme.tok_id().try_into().ok().unwrap();
+                grm.term_name(TIdx::from(token_id)).unwrap().to_string()
+            }
+        }
+    };
+
+    fn gen_class_body(node: &Node<u16>, grm: &YaccGrammar, input: &str, class: &mut Class) {
+        if let &Node::Nonterm { nonterm_idx, ref nodes } = node {
+            for child in nodes.iter() {
+                if let &Node::Nonterm { nonterm_idx, ref nodes} = child {
+                    let name = term_name(child, grm, input);
+                    match name.as_ref() {
+                        "method_def" => gen_method(child, grm, input, class),
+                        "class_body" => gen_class_body(child, grm, input, class),
+                        _ => (),
+                    }
+                }
+            }
+        }
+    }
+
+    // method_def : "DEF" "IDENTIFIER" "LPAREN" parameter_list_opt "RPAREN" block ;
+    fn gen_method(node: &Node<u16>, grm: &YaccGrammar, input: &str, class: &mut Class) {
+        if let &Node::Nonterm{ nonterm_idx, ref nodes } = node {
+            let mut method = Method::new();
```

Comment:
I'm not a big fan of building `Method` up in this way. Wouldn't it be clearer if things like `gen_block` returned a code block, rather than stuffing it inside something? This is doubly so, because you might want `gen_block` to be used in non-method contexts.

Reply:
Makes sense,  I think I was overthinking it. gen_block now adds all bytecode instrs to the class block it is inside, along with a symbol table of methods which provides useful static information for the interpreter (var names, num of args etc).